File: PersonalFinanceManagement/Services/ExportService.cs
using System;
using System.IO;
using System.Threading.Tasks;
using System.Linq;
using System.Collections.Generic;
using Microsoft.Win32;
using ClosedXML.Excel;
using PersonalFinanceManagement.Models;

namespace PersonalFinanceManagement.Services
{
    public interface IExportService
    {
        Task ExportToExcelAsync(
            IEnumerable<Income> incomes,
            IEnumerable<Expense> expenses,
            DateTime startDate,
            DateTime endDate);
    }

    public class ExportService : IExportService
    {
        public async Task ExportToExcelAsync(
            IEnumerable<Income> incomes,
            IEnumerable<Expense> expenses,
            DateTime startDate,
            DateTime endDate)
        {
            var dialog = new SaveFileDialog
            {
                Filter = "Excel Files|*.xlsx",
                DefaultExt = ".xlsx",
                FileName = $"Báo_Cáo_Tài_Chính_{startDate:dd-MM-yyyy}_đến_{endDate:dd-MM-yyyy}"
            };

            if (dialog.ShowDialog() == true)
            {
                await Task.Run(() =>
                {
                    using var workbook = new XLWorkbook();
                    
                    // Create Income Sheet
                    var incomeSheet = workbook.Worksheets.Add("Thu nhập");
                    AddIncomeData(incomeSheet, incomes);

                    // Create Expense Sheet
                    var expenseSheet = workbook.Worksheets.Add("Chi tiêu");
                    AddExpenseData(expenseSheet, expenses);

                    // Create Summary Sheet
                    var summarySheet = workbook.Worksheets.Add("Tổng kết");
                    AddSummaryData(summarySheet, incomes, expenses);

                    workbook.SaveAs(dialog.FileName);
                });
            }
        }

        private void AddIncomeData(IXLWorksheet sheet, IEnumerable<Income> incomes)
        {
            // Headers
            sheet.Cell(1, 1).Value = "Ngày";
            sheet.Cell(1, 2).Value = "Số tiền (VNĐ)";
            sheet.Cell(1, 3).Value = "Danh mục";
            sheet.Cell(1, 4).Value = "Nguồn thu";
            sheet.Cell(1, 5).Value = "Ghi chú";

            // Style headers
            var headerRow = sheet.Row(1);
            headerRow.Style.Font.Bold = true;
            headerRow.Style.Fill.BackgroundColor = XLColor.LightGray;

            // Data
            int row = 2;
            foreach (var income in incomes.OrderBy(i => i.Date))
            {
                sheet.Cell(row, 1).Value = income.Date;
                sheet.Cell(row, 2).Value = income.Amount;
                sheet.Cell(row, 3).Value = income.Category?.Name;
                sheet.Cell(row, 4).Value = income.Source;
                sheet.Cell(row, 5).Value = income.Description;
                row++;
            }

            // Format
            sheet.Column(1).Style.DateFormat.Format = "dd/MM/yyyy";
            sheet.Column(2).Style.NumberFormat.Format = "#,##0";
            sheet.Columns().AdjustToContents();
        }

        private void AddExpenseData(IXLWorksheet sheet, IEnumerable<Expense> expenses)
        {
            // Headers
            sheet.Cell(1, 1).Value = "Ngày";
            sheet.Cell(1, 2).Value = "Số tiền (VNĐ)";
            sheet.Cell(1, 3).Value = "Danh mục";
            sheet.Cell(1, 4).Value = "Phương thức";
            sheet.Cell(1, 5).Value = "Thiết yếu";
            sheet.Cell(1, 6).Value = "Ghi chú";

            // Style headers
            var headerRow = sheet.Row(1);
            headerRow.Style.Font.Bold = true;
            headerRow.Style.Fill.BackgroundColor = XLColor.LightGray;

            // Data
            int row = 2;
            foreach (var expense in expenses.OrderBy(e => e.Date))
            {
                sheet.Cell(row, 1).Value = expense.Date;
                sheet.Cell(row, 2).Value = expense.Amount;
                sheet.Cell(row, 3).Value = expense.Category?.Name;
                sheet.Cell(row, 4).Value = expense.PaymentMethod;
                sheet.Cell(row, 5).Value = expense.IsEssential ? "Có" : "Không";
                sheet.Cell(row, 6).Value = expense.Description;
                row++;
            }

            // Format
            sheet.Column(1).Style.DateFormat.Format = "dd/MM/yyyy";
            sheet.Column(2).Style.NumberFormat.Format = "#,##0";
            sheet.Columns().AdjustToContents();
        }

        private void AddSummaryData(IXLWorksheet sheet, IEnumerable<Income> incomes, IEnumerable<Expense> expenses)
        {
            // Title
            sheet.Cell(1, 1).Value = "Báo Cáo Tổng Kết";
            sheet.Range(1, 1, 1, 2).Merge().Style.Font.Bold = true;

            // Income Summary
            sheet.Cell(3, 1).Value = "Tổng thu nhập:";
            sheet.Cell(3, 2).Value = incomes.Sum(i => i.Amount);

            // Expense Summary
            sheet.Cell(4, 1).Value = "Tổng chi tiêu:";
            sheet.Cell(4, 2).Value = expenses.Sum(e => e.Amount);

            // Balance
            sheet.Cell(5, 1).Value = "Số dư:";
            sheet.Cell(5, 2).Value = incomes.Sum(i => i.Amount) - expenses.Sum(e => e.Amount);

            // Essential vs Non-Essential Expenses
            sheet.Cell(7, 1).Value = "Chi tiêu thiết yếu:";
            sheet.Cell(7, 2).Value = expenses.Where(e => e.IsEssential).Sum(e => e.Amount);

            sheet.Cell(8, 1).Value = "Chi tiêu không thiết yếu:";
            sheet.Cell(8, 2).Value = expenses.Where(e => !e.IsEssential).Sum(e => e.Amount);

            // Format
            sheet.Column(2).Style.NumberFormat.Format = "#,##0";
            sheet.Columns().AdjustToContents();

            // Style
            var range = sheet.Range(3, 1, 8, 2);
            range.Style.Border.OutsideBorder = XLBorderStyleValues.Thin;
            range.Style.Border.InsideBorder = XLBorderStyleValues.Thin;
        }
    }
}

File: apps/services/export_service.py
import openpyxl
from openpyxl.styles import Font, PatternFill, Alignment, Border, Side
from typing import List, Optional
from pathlib import Path
from datetime import datetime
from apps.models.transaction import Income, Expense

class ExportService:
    @classmethod
    def export_to_excel(
        cls,
        incomes: List[Income],
        expenses: List[Expense],
        start_date: str,
        end_date: str,
        output_path: Optional[Path] = None
    ) -> Path:
        """
        Export Incomes, Expenses, and Financial Summary to Excel (.xlsx) matching WPF ClosedXML structure.
        """
        if output_path is None:
            filename = f"Bao_Cao_Tai_Chinh_{start_date}_den_{end_date}.xlsx"
            output_dir = Path.home() / "Downloads"
            if not output_dir.exists():
                output_dir = Path("./data")
                output_dir.mkdir(exist_ok=True)
            output_path = output_dir / filename

        wb = openpyxl.Workbook()
        
        # Styles
        header_font = Font(name="Arial", size=11, bold=True, color="000000")
        header_fill = PatternFill(start_color="E2E8F0", end_color="E2E8F0", fill_type="solid")
        thin_border = Border(
            left=Side(style="thin", color="CBD5E1"),
            right=Side(style="thin", color="CBD5E1"),
            top=Side(style="thin", color="CBD5E1"),
            bottom=Side(style="thin", color="CBD5E1"),
        )
        title_font = Font(name="Arial", size=14, bold=True, color="1E293B")
        summary_header_font = Font(name="Arial", size=11, bold=True)

        # 1. Sheet Thu nhập
        ws_income = wb.active
        ws_income.title = "Thu nhập"
        ws_income.views.sheetView[0].showGridLines = True

        income_headers = ["Ngày", "Số tiền (VNĐ)", "Danh mục", "Nguồn thu", "Ghi chú"]
        ws_income.append(income_headers)
        for col in range(1, len(income_headers) + 1):
            cell = ws_income.cell(row=1, column=col)
            cell.font = header_font
            cell.fill = header_fill
            cell.alignment = Alignment(horizontal="center", vertical="center")

        for inc in sorted(incomes, key=lambda x: x.date):
            ws_income.append([
                inc.date,
                inc.amount,
                inc.category.name if inc.category else "Chưa phân loại",
                inc.source,
                inc.description
            ])

        for row in range(2, ws_income.max_row + 1):
            ws_income.cell(row=row, column=2).number_format = "#,##0"

        cls._auto_fit_columns(ws_income)

        # 2. Sheet Chi tiêu
        ws_expense = wb.create_sheet(title="Chi tiêu")
        ws_expense.views.sheetView[0].showGridLines = True

        expense_headers = ["Ngày", "Số tiền (VNĐ)", "Danh mục", "Phương thức", "Thiết yếu", "Ghi chú"]
        ws_expense.append(expense_headers)
        for col in range(1, len(expense_headers) + 1):
            cell = ws_expense.cell(row=1, column=col)
            cell.font = header_font
            cell.fill = header_fill
            cell.alignment = Alignment(horizontal="center", vertical="center")

        for exp in sorted(expenses, key=lambda x: x.date):
            ws_expense.append([
                exp.date,
                exp.amount,
                exp.category.name if exp.category else "Chưa phân loại",
                exp.payment_method,
                "Có" if exp.is_essential else "Không",
                exp.description
            ])

        for row in range(2, ws_expense.max_row + 1):
            ws_expense.cell(row=row, column=2).number_format = "#,##0"

        cls._auto_fit_columns(ws_expense)

        # 3. Sheet Tổng kết
        ws_summary = wb.create_sheet(title="Tổng kết")
        ws_summary.views.sheetView[0].showGridLines = True

        ws_summary.cell(row=1, column=1, value="Báo Cáo Tổng Kết Tài Chính").font = title_font
        ws_summary.cell(row=2, column=1, value=f"Kỳ báo cáo: {start_date} đến {end_date}")

        total_income = sum(i.amount for i in incomes)
        total_expense = sum(e.amount for e in expenses)
        balance = total_income - total_expense
        essential_expense = sum(e.amount for e in expenses if e.is_essential)
        non_essential_expense = sum(e.amount for e in expenses if not e.is_essential)

        summary_data = [
            ("Tổng thu nhập:", total_income),
            ("Tổng chi tiêu:", total_expense),
            ("Số dư:", balance),
            ("Chi tiêu thiết yếu:", essential_expense),
            ("Chi tiêu không thiết yếu:", non_essential_expense),
        ]

        start_row = 4
        for idx, (label, val) in enumerate(summary_data):
            curr_row = start_row + idx
            lbl_cell = ws_summary.cell(row=curr_row, column=1, value=label)
            val_cell = ws_summary.cell(row=curr_row, column=2, value=val)

            lbl_cell.font = summary_header_font
            lbl_cell.border = thin_border
            val_cell.font = summary_header_font
            val_cell.number_format = "#,##0"
            val_cell.border = thin_border

        cls._auto_fit_columns(ws_summary)

        wb.save(str(output_path))
        return output_path

    @staticmethod
    def _auto_fit_columns(ws):
        for col in ws.columns:
            max_len = 0
            col_letter = col[0].column_letter
            for cell in col:
                val_str = str(cell.value or '')
                max_len = max(max_len, len(val_str))
            ws.column_dimensions[col_letter].width = max(max_len + 4, 12)
